{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e91b3f51",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import mygene\n",
    "import statistics\n",
    "import glob\n",
    "os.chdir(\"../../data/recipients/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af410a8d-f8ad-46ce-8da6-198f41217129",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ensembl2gene():\n",
    "    ensmeble2gene={}\n",
    "    fx=pd.read_csv(\"wgcna/gene2ensemble_fx.csv\")\n",
    "    hx=pd.read_csv(\"wgcna/gene2ensemble_hx.csv\")\n",
    "    temp=pd.concat([fx,hx])\n",
    "    genes={}\n",
    "    for i,row in temp.iterrows():\n",
    "        genes[row[\"ensbl\"]]=row[\"gene\"]\n",
    "    \n",
    "    return genes\n",
    "\n",
    "\n",
    "def intersection(lst1, lst2):\n",
    "    lst3 = [value for value in lst1 if value in lst2]\n",
    "    return lst3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5b4eb23d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "edges=pd.read_csv(\"mmvec.edge.recipients.txt\",sep=\"\\t\")\n",
    "nodes=pd.read_csv(\"wKDA_reanalysis/q6JZ4or4TB.wKDA_cytoscape_nodes.txt\",sep=\"\\t\")\n",
    "kds=pd.read_csv(\"wKDA_reanalysis/q6JZ4or4TB.wKDA_kd_full_results.txt\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c80f4780",
   "metadata": {},
   "outputs": [],
   "source": [
    "degs_hx=[]\n",
    "degs_fx=[]\n",
    "for i,row in nodes[nodes['MODULE'].notna()].iterrows():\n",
    "    module=row[\"MODULE\"]\n",
    "    node=row[\"NODE\"]\n",
    "    if(\"DEG_fx\" in module):\n",
    "        degs_hx.append(node.replace(\"_hx\",\"\"))\n",
    "    elif(\"DEG_hx\" in module):\n",
    "        degs_fx.append(node.replace(\"_fx\",\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2a8071e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "en2gene=ensembl2gene()\n",
    "degs_hx=[]\n",
    "for i in en2gene.keys():\n",
    "    degs_hx.append(en2gene[i]+\"_hx\")\n",
    "\n",
    "en2gene=ensembl2gene()\n",
    "degs_fx=[]\n",
    "for i in en2gene.keys():\n",
    "    degs_fx.append(en2gene[i]+\"_fx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c277f1d8-856c-4d33-bdf5-121e27aef436",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwasoverlap=pd.read_csv(\"wgcna/GWAS.overlap.txt\",sep=\"\\t\")\n",
    "gwas_overlap=list(gwasoverlap[\"MGI.symbol\"])\n",
    "gwas_overlap=[]\n",
    "for gene in gwasoverlap[\"MGI.symbol\"]:\n",
    "    gwas_overlap.append(gene+\"_hx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2aa3e9e4-1a31-4b16-9f03-a10a138a6a40",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes=ensembl2gene()\n",
    "\n",
    "for i,row in edges.iterrows():\n",
    "    if \"hx\" in row[\"HEAD\"]:\n",
    "        gene=row[\"HEAD\"].split(\"_\")[0]\n",
    "        if(gene in genes.keys()):\n",
    "            edges.iloc[i,0]=genes[gene]+\"_hx\"\n",
    "    elif \"fx\" in row[\"HEAD\"]:\n",
    "        gene=row[\"HEAD\"].split(\"_\")[0]\n",
    "        if(gene in genes.keys()):\n",
    "            edges.iloc[i,0]=genes[gene]+\"_fx\"\n",
    "            \n",
    "    if \"hx\" in row[\"TAIL\"]:\n",
    "        gene=row[\"TAIL\"].split(\"_\")[0]\n",
    "        if(gene in genes.keys()):\n",
    "            edges.iloc[i,1]=genes[gene]+\"_hx\"\n",
    "    elif \"fx\" in row[\"TAIL\"]:\n",
    "        gene=row[\"TAIL\"].split(\"_\")[0]\n",
    "        if(gene in genes.keys()):\n",
    "            edges.iloc[i,1]=genes[gene]+\"_fx\"\n",
    "\n",
    "for i,row in nodes.iterrows():\n",
    "    if \"hx\" in row[\"NODE\"]:\n",
    "        gene=row[\"NODE\"].split(\"_\")[0]\n",
    "        if(gene in genes.keys()):\n",
    "            nodes.iloc[i,0]=genes[gene]+\"_hx\"\n",
    "            nodes.iloc[i,1]=genes[gene]+\"_hx\"\n",
    "    if \"fx\" in row[\"NODE\"]:\n",
    "        gene=row[\"NODE\"].split(\"_\")[0]\n",
    "        if(gene in genes.keys()):\n",
    "            nodes.iloc[i,0]=genes[gene]+\"_fx\"\n",
    "            nodes.iloc[i,1]=genes[gene]+\"_fx\"\n",
    "\n",
    "for i,row in kds.iterrows(): \n",
    "    if \"hx\" in row[\"NODE\"]:\n",
    "        gene=row[\"NODE\"].split(\"_\")[0]\n",
    "        if(gene in genes.keys()):\n",
    "            kds.iloc[i,1]=genes[gene]+\"_hx\"\n",
    "    if \"fx\" in row[\"NODE\"]:\n",
    "        gene=row[\"NODE\"].split(\"_\")[0]\n",
    "        if(gene in genes.keys()):\n",
    "            kds.iloc[i,1]=genes[gene]+\"_fx\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82e9dee8-d725-4ddb-8bef-3d6210acb620",
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes[\"GWAS\"]=\"\"\n",
    "nodes[\"DEGS\"]=\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a4c20ba-a2fa-4f3e-9979-f0afb78ba17f",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_kds=[\"DTDPRHAMSYN-PWY: dTDP-L-rhamnose biosynthesis I|g__Bacteroides.s__Bacteroides_cellulosilyticus\",\n",
    "             \"Atf4_hx\",\"PWY-5101: L-isoleucine biosynthesis II|g__Lachnoclostridium.s__Clostridium_hylemonae\",\n",
    "             \"PWY-6507: 4-deoxy-L-threo-hex-4-enopyranuronate degradation|g__Lachnoclostridium.s__Clostridium_aldenense\",\n",
    "             \"Olfm1_hx\",\n",
    "             \"Gm15206_fx\",\n",
    "             \"Zfp385b_hx\",\n",
    "             \"Tor1b_hx\",\n",
    "             \"Gm37848_hx\",\n",
    "              \"Lcmt2_hx\",\n",
    "              \"Stra6_fx\",\n",
    "              \"PWY-7446: sulfoglycolysis|g__Citrobacter.s__Citrobacter_amalonaticus\",\n",
    "              \"Cxcl10_hx\",\n",
    "              \"Mrps2_hx\",\n",
    "              \"Loxl3_hx\",\n",
    "              \"Rassf6_hx\",\n",
    "              \"PWY-1861: formaldehyde assimilation II (RuMP Cycle)|g__Lachnoclostridium.s__Clostridium_aldenense\",\n",
    "              \"PWY-6609: adenine and adenosine salvage III|g__Bacteroides.s__Bacteroides_sartorii\",\n",
    "              \"VALSYN-PWY: L-valine biosynthesis|g__Bacteroides.s__Bacteroides_sartorii\",\n",
    "              \"PWY-6703: preQ0 biosynthesis|g__Alistipes.s__Alistipes_putredinis\",\n",
    "              \"PWY-3841: folate transformations II|g__Bacteroides.s__Bacteroides_xylanisolvens\"\n",
    "             ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0eb5e8cf-16a7-4191-a66a-1921a7ab146a",
   "metadata": {},
   "outputs": [],
   "source": [
    "node2module=pd.read_csv(\"wgcna/NODES4wKDA.tsv\",sep=\"\\t\")\n",
    "node2module=node2module.set_index(\"NODE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70aa592a-bc39-4d8d-a875-b3dca9aee400",
   "metadata": {},
   "outputs": [],
   "source": [
    "modulepath={\"hx\":\"../../modules/modules-hx/gene_name/\",\n",
    "           \"fx\":\"../../modules/modules-fx/gene_name/\",\n",
    "           \"serum\":\"../../modules/modules-serum/modules/\",\n",
    "           \"feces\":\"../../modules/modules-feces/modules/\",\n",
    "           \"meta\":\"../../modules/metagenome_modules/\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc41fb80-357d-4fa6-81ec-e62ce078be89",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes=ensembl2gene()\n",
    "\n",
    "for i,row in node2module.iterrows():\n",
    "    if \"hx\" in row[\"NODE\"]:\n",
    "        gene=row[\"NODE\"].split(\"_\")[0]\n",
    "        if(gene in genes.keys()):\n",
    "            node2module.iloc[i,1]=genes[gene]+\"_hx\"\n",
    "    if \"fx\" in row[\"NODE\"]:\n",
    "        gene=row[\"NODE\"].split(\"_\")[0]\n",
    "        if(gene in genes.keys()):\n",
    "            node2module.iloc[i,1]=genes[gene]+\"_fx\"\n",
    "for tissue in modulepath.keys():\n",
    "    mp=modulepath[tissue]\n",
    "    module_files=glob.glob(mp+\"/*.txt\")\n",
    "    for mf in module_files:\n",
    "        md=os.path.basename(mf).replace(\".txt\",\"\").replace(\".new\",\"\")\n",
    "        with open(mf,\"r\") as f:\n",
    "            for line in f:\n",
    "                if(tissue ==\"meta\"):\n",
    "                    node2module=node2module.append({\"MODULE\":md+\"_\"+tissue,\"NODE\":line.strip()},ignore_index = True)\n",
    "                    #node2module[line.strip()]=md+\"_\"+tissue\n",
    "                elif(tissue==\"fx\" or tissue==\"hx\"):\n",
    "                    node2module=node2module.append({\"MODULE\":md+\"_\"+tissue,\"NODE\":line.strip().capitalize()+\"_\"+tissue},ignore_index = True)\n",
    "                else:\n",
    "                    node2module=node2module.append({\"MODULE\":md+\"_\"+tissue,\"NODE\":line.strip()+\"_\"+tissue},ignore_index = True)\n",
    "                    #node2module[line.strip()+\"_\"+tissue]=md+\"_\"+tissue\n",
    "node2module=node2module.drop_duplicates()\n",
    "node2module=node2module.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c0d73d3-4eaf-41be-a540-1886ae96ab34",
   "metadata": {},
   "outputs": [],
   "source": [
    "deg_fx=pd.read_csv(\"wgcna/DEGs/Degs_frontal_cortex.txt\",sep=\"\\t\")+\"_fx\"\n",
    "deg_hx=pd.read_csv(\"wgcna/DEGs/Degs_hippo.txt\",sep=\"\\t\")+\"_hx\"\n",
    "deg_feces=pd.read_csv(\"wgcna/DEGs/degs_feces.txt\",sep=\"\\t\")\n",
    "deg_serum=pd.read_csv(\"wgcna/DEGs/degs_serum.txt\",sep=\"\\t\")\n",
    "deg_fx.columns=[\"NODE\"]\n",
    "deg_hx.columns=[\"NODE\"]\n",
    "deg=pd.concat([deg_fx,deg_hx,deg_feces,deg_serum])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45063406-1f5f-4923-a9f1-281d6317ac95",
   "metadata": {},
   "outputs": [],
   "source": [
    "node2module=node2module.set_index(\"MODULE\")\n",
    "deg_gwas_hx=list(node2module.loc[\"GWAS_hx\",\"NODE\"])\n",
    "node2module=node2module.reset_index()\n",
    "node2module=node2module.set_index(\"NODE\")\n",
    "kds=kds.reset_index()\n",
    "kds=kds.set_index(\"NODE\")\n",
    "node2module=node2module.reset_index()\n",
    "node2module=node2module.set_index(\"MODULE\")\n",
    "node2module=node2module.loc[list(set(node2module.index)-set([\"DEG_fx\",\"DEG_hx\",\"DEG_serum\",\"DEG_feces\",\"GWAS_hx\"])),:]\n",
    "node2module=node2module.reset_index()\n",
    "node2module=node2module.set_index(\"NODE\")\n",
    "nodes=nodes.set_index(\"NODE\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec2d140e-6865-4c37-b878-08c9149972ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "module4kd_count={}\n",
    "for kd in selected_kds:\n",
    "    moduleFromKDS=[]\n",
    "    moduleFromnode2module=[]\n",
    "    if(isinstance(kds.loc[kd,\"MODULE\"], pd.core.series.Series)):\n",
    "        tmplist=[]\n",
    "        for md in kds.loc[kd,\"MODULE\"].tolist():\n",
    "            tmplist.append(md+\"_\"+kd)\n",
    "        moduleFromKDS=tmplist\n",
    "        #moduleFromKDS=kds.loc[kd,\"MODULE\"].tolist()\n",
    "    else:\n",
    "        moduleFromKDS=[kds.loc[kd,\"MODULE\"]+\"_\"+kd]\n",
    "    if(isinstance(node2module.loc[kd,\"MODULE\"], pd.core.series.Series)):\n",
    "        tmplist=[]\n",
    "        for md in node2module.loc[kd,\"MODULE\"].tolist():\n",
    "            tmplist.append(md+\"_\"+kd)\n",
    "        moduleFromKDS=tmplist\n",
    "        #moduleFromnode2module=node2module.loc[kd,\"MODULE\"].tolist()\n",
    "    else:\n",
    "        moduleFromnode2module=[node2module.loc[kd,\"MODULE\"]+\"_\"+kd]\n",
    "    modules4kd=set(moduleFromKDS+ moduleFromnode2module)\n",
    "    for md in modules4kd:\n",
    "        module4kd_count[md]=0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dae10af8-a737-4d1c-85a8-990a2dbcde48",
   "metadata": {},
   "outputs": [],
   "source": [
    "kd_list=[]\n",
    "module4kd_count={}\n",
    "for kd in selected_kds:\n",
    "    moduleFromKDS=[]\n",
    "    moduleFromnode2module=[]\n",
    "    if(isinstance(kds.loc[kd,\"MODULE\"], pd.core.series.Series)):\n",
    "        tmplist=[]\n",
    "        for md in kds.loc[kd,\"MODULE\"].tolist():\n",
    "            tmplist.append(md+\"_\"+kd)\n",
    "        moduleFromKDS=tmplist\n",
    "        #moduleFromKDS=kds.loc[kd,\"MODULE\"].tolist()\n",
    "    else:\n",
    "        moduleFromKDS=[kds.loc[kd,\"MODULE\"]+\"_\"+kd]\n",
    "    if(isinstance(node2module.loc[kd,\"MODULE\"], pd.core.series.Series)):\n",
    "        tmplist=[]\n",
    "        for md in node2module.loc[kd,\"MODULE\"].tolist():\n",
    "            tmplist.append(md+\"_\"+kd)\n",
    "        moduleFromKDS=tmplist\n",
    "        #moduleFromnode2module=node2module.loc[kd,\"MODULE\"].tolist()\n",
    "    else:\n",
    "        moduleFromnode2module=[node2module.loc[kd,\"MODULE\"]+\"_\"+kd]\n",
    "    modules4kd=set(moduleFromKDS+ moduleFromnode2module)\n",
    "    for md in modules4kd:\n",
    "        module4kd_count[md]=0\n",
    "\n",
    "for kd in selected_kds:\n",
    "    subset=edges.loc[edges[\"HEAD\"]==kd,:]\n",
    "    kd_on_tail=False\n",
    "    if len(subset) ==0:\n",
    "        subset=edges.loc[edges[\"TAIL\"]==kd,:]\n",
    "        kd_on_tail=True\n",
    "    for i,row in subset.iterrows():\n",
    "        target=row[\"TAIL\"]\n",
    "        if(row[\"TAIL\"]==row[\"HEAD\"]):\n",
    "            continue\n",
    "        if(kd_on_tail):\n",
    "            target=row[\"HEAD\"]\n",
    "#         if \"_hx\" in target or \"_fx\" in target:\n",
    "#             print(target)\n",
    "            #target=target.capitalize()\n",
    "        node_module=node2module[node2module.index==target][\"MODULE\"]\n",
    "        #node_module=node2module.loc[target,\"MODULE\"]\n",
    "        #if in deg list and module\n",
    "        #SCAN For DEGS and GWAS first\n",
    "        if(isinstance(node_module, pd.core.series.Series)):\n",
    "            for nm in node_module:\n",
    "                nm=nm+\"_\"+kd\n",
    "#                 if(nm not in module4kd_count.keys()):\n",
    "#                     module4kd_count[nm]=0\n",
    "                if(nm in module4kd_count.keys()):\n",
    "                    if(target.casefold() in (node.casefold() for node in deg[\"NODE\"])):\n",
    "                        kd_list.append(row.to_frame().transpose())\n",
    "                        module4kd_count[nm]=module4kd_count[nm]+1\n",
    "                        nodes.loc[target,\"DEGS\"]=\"T\"\n",
    "                    if(target in deg_gwas_hx):\n",
    "                        kd_list.append(row.to_frame().transpose())\n",
    "                        module4kd_count[nm]=module4kd_count[nm]+1\n",
    "                        nodes.loc[target,\"GWAS\"]=\"T\"\n",
    "                    if(\"_serum\" in target):\n",
    "                        print(target)\n",
    "                        kd_list.append(row.to_frame().transpose())\n",
    "                        module4kd_count[nm]=module4kd_count[nm]+1\n",
    "                    \n",
    "        else:\n",
    "            node_module=node_module+\"_\"+kd\n",
    "#             if(node_module not in module4kd_count.keys()):\n",
    "#                 module4kd_count[node_module]=0\n",
    "            if(node_module in module4kd_count.keys()):\n",
    "                if(target.casefold() in (node.casefold() for node in deg[\"NODE\"])):\n",
    "                    kd_list.append(row.to_frame().transpose())\n",
    "                    module4kd_count[node_module]=module4kd_count[node_module]+1\n",
    "                    nodes.loc[target,\"DEGS\"]=\"T\"\n",
    "                if(target in deg_gwas_hx):\n",
    "                    kd_list.append(row.to_frame().transpose())\n",
    "                    module4kd_count[node_module]=module4kd_count[node_module]+1\n",
    "                    nodes.loc[target,\"GWAS\"]=\"T\"\n",
    "                if(\"_serum\" in target):\n",
    "                        print(target)\n",
    "                        kd_list.append(row.to_frame().transpose())\n",
    "                        module4kd_count[nm]=module4kd_count[nm]+1\n",
    "                    \n",
    "    #Fill rest of the network connected to KD \n",
    "    for i,row in subset.iterrows():\n",
    "        target=row[\"TAIL\"]\n",
    "        if(row[\"TAIL\"]==row[\"HEAD\"]):\n",
    "            continue\n",
    "        if(kd_on_tail):\n",
    "            target=row[\"HEAD\"]\n",
    "        if \"_hx\" in target or \"_fx\" in target:\n",
    "            target=target.capitalize()\n",
    "        node_module=node2module[node2module.index==target][\"MODULE\"]\n",
    "        #node_module=node2module.loc[target,\"MODULE\"]\n",
    "        #if in deg list and module\n",
    "        if(isinstance(node_module, pd.core.series.Series)):\n",
    "            for nm in node_module:\n",
    "                nm=nm+\"_\"+kd\n",
    "                if(nm in module4kd_count.keys()):\n",
    "                    if module4kd_count[nm] < 10:\n",
    "                        kd_list.append(row.to_frame().transpose())\n",
    "                        module4kd_count[nm]=module4kd_count[nm]+1\n",
    "        else:\n",
    "            node_module=node_module+\"_\"+kd\n",
    "            if(node_module in module4kd_count.keys()):\n",
    "                if module4kd_count[node_module] < 10:\n",
    "                    kd_list.append(row.to_frame().transpose())\n",
    "                    module4kd_count[node_module]=module4kd_count[node_module]+1\n",
    "                    \n",
    "nodes.to_csv(\"newnodes.txt\",sep=\"\\t\",index=False)\n",
    "selected_edges=pd.concat(kd_list).reset_index(drop=True).drop_duplicates()\n",
    "selected_edges.to_csv(\"selected_edges.txt\",sep=\"\\t\",index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
